v2/internal/controller/controllers/targetnamespace_test.go

```diff
@@ -0,0 +1,258 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package controllers_test
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	resources "github.com/Azure/azure-service-operator/v2/api/microsoft.resources/v1alpha1api20200601"
+	"github.com/Azure/azure-service-operator/v2/internal/controller/controllers"
+	"github.com/Azure/azure-service-operator/v2/internal/controller/reconcilers"
+	"github.com/Azure/azure-service-operator/v2/internal/controller/testcommon"
+
+	. "github.com/onsi/gomega"
+)
+
+const finalizerName = reconcilers.GenericControllerFinalizer
+
+const (
+	timeoutFast = time.Minute * 3
+	retry       = time.Second * 3
+)
+
+func TestTargetNamespaces(t *testing.T) {
+	t.Parallel()
+	tc := globalTestContext.ForTest(t)
+
+	createNamespaces(tc, "watched", "unwatched")
+	configuredNamespaces := os.Getenv("AZURE_TARGET_NAMESPACES")
+	podNamespace := os.Getenv("POD_NAMESPACE")
+
+	// We can't check for operator namespace if there's not one set.
+	tc.Expect(podNamespace).ToNot(Equal(""), "no POD_NAMESPACE set")
+
+	standardSpec := resources.ResourceGroupSpec{
+		Location: tc.AzureRegion,
+		Tags:     testcommon.CreateTestResourceGroupDefaultTags(),
+	}
+	// Create resource groups in these two namespaces - we can't
+	// easily use the test context for this because that's geared to
+	// creating things in one test namespace, while we need fixed
+	// namespaces.
+	rgDefault := resources.ResourceGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      tc.Namer.GenerateName("rg"),
+			Namespace: "default",
+		},
+		Spec: standardSpec,
+	}
+	tc.CreateResourceGroupAndWait(&rgDefault)
+	// Check that the instance is annotated with the operator namespace.
+	checkNamespaceAnnotation(tc, &rgDefault, podNamespace)
+
+	// The watched namespace is also reconciled.
+	rgWatched := resources.ResourceGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      tc.Namer.GenerateName("rg"),
+			Namespace: "watched",
+		},
+		Spec: standardSpec,
+	}
+	tc.CreateResourceGroupAndWait(&rgWatched)
+	checkNamespaceAnnotation(tc, &rgWatched, podNamespace)
+
+	// But the unwatched namespace isn't...
+	unwatchedName := tc.Namer.GenerateName("rg")
+	rgUnwatched := resources.ResourceGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      unwatchedName,
+			Namespace: "unwatched",
```

Comment:
minor: maybe `unwatchedNamespace` variable to go along with `unwatchedName`?

Reply:
Or just construct the `NamespacedName` from below earlier and then use that everywhere.